### ISSUE TITLE
Empty statements lint should not count for/while loops

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/linter#383]
For example, this is entirely kosher (you could be iterating through something):

```dart
while (something.meetsCondition());
```

**Comment by alexeieleusis** [comment]
I think in this case having the ignore annotation is better, it conveys maintainers that it is intentional.

Unless `something` is in a third party library, I think it should have a `getInUsableCondition()` method or similar, I would expect this kind of occurrences be pretty rare.

**Comment by matanlurey** [comment]
I don't agree, you shouldn't have to ignore valid code that contains a valid common pattern.

Almost every parser-type library will do something like this.

**Comment by zoechi** [comment]
>  you shouldn't have to ignore valid code 

In this case the `// ignore: ...` feature would be useless

> valid common pattern.

Also `unawaited_futures` are quite common but it still causes a linter warning.

I guess this requires some generic discussion how much possible missed bugs weight against the inconvenience of suppressing hints for valid code.

**Comment by matanlurey** [comment]
I strongly disagree with `unawaited_futures` even existing (and I won't allow it on any of my projects or official dart-lang ones) - there is the difference between "suggest possible bugs for me to review" and "force me to ignore totally valid code".

**Comment by zoechi** [comment]
I don't have a strong opinion on ether side. I just wish for some consistency.
Perhaps splitting it into two rules for sure and possible bugs?

**Comment by alexeieleusis** [comment]
I think that's the idea behind having lints and hints, hints cannot be
opted out and lints are opt in.

Alexei Diaz | Software Engineer | alexeidiaz@google.com | (650) 214-6432


On Wed, Jan 25, 2017 at 10:36 PM, Günter Zöchbauer <notifications@github.com

**Comment by matanlurey** [comment]
At least as defined by Google, lints are something you warn about *every time* you see them. If that's the case maybe we need something like a "hint-lint" that you can completely disable (I'd never like to see them).

**Comment by bwilkerson** [comment]
Thank you, but we don't need more categories of diagnostics. :-)

Hints are for cases that are clearly bugs in code but that the spec doesn't define to be an error or warning. Hints must have zero false positives.

Lints are for cases that are more stylistic in nature, where some people will want to follow a rule and others won't. Lints should have zero false positives, but in some rare cases where it's too expensive to remove the false positives and they are extremely rare we *might* choose to allow them. In those cases it is possible to locally disable the lints where they would generate a false positive. (Hence the motivation for that feature.)

As @zoechi suggested, if we have a lint that some people want to follow in some cases but not others, we should seriously consider splitting the rule. People who want both rules to be in effect can enable both lints.

**Comment by matanlurey** [comment]
I'd be a fan of splitting the lint

**Comment by pq** [comment]
FWIW: if splitting the lint means more folks get value from it, I'm all for it.

We've strayed a bit from the original reported issue though.  Should we take this to a fresh enhancement request?

As for @matanlurey's observation that this triggers a lint:

```dart
while (something.meetsCondition());
```

I think that's pretty common for this lint as implemented elsewhere.  An escape hatch might be to notice if there's a comment in the body.

```dart
while (something.meetsCondition()) {
  // iterate, really
}
```

(No lint here.)

This is, for example, what eslint [does](http://eslint.org/docs/rules/no-empty) (not that this is an exemplar necessarily!)

**Comment by matanlurey** [comment]
I no longer feel passionate about this issue :)
